sbomtool merge and filter features

rpkelly opened the pull request:
<!--
Tips:
- Please read the CONTRIBUTING document to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
#296


**Description of changes:**
Adds the following features to `sbomtool`:

* Filtering command that removes information not relevant to the actually built image
* Merge command to merge multiple sboms
* Integration test suite using godog
* Deduplication feature, to remove packages added multiple times to the same sbom


**Testing done:**
`make all`


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

rpkelly committed:
chore: add makefile and gitignore to sbomtool

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

rpkelly committed:
feat: add merge sboms feature

sbomtool merge feature allows multiple sboms of the same type to be
merged into a single sbom. This change allows us to merge the sboms of
multiple individual packages into a large sbom for an image.

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

rpkelly committed:
feat: add deduplication package for SBOM processing

When generating a large SBOM, and especially when merging multiple SBOMs
together, it is likely that the same package will be found multiple
times. The deduplication code feature replaces those duplicates with a
single merged package, and reroutes all existing relationships to the
new merged package

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

rpkelly committed:
refactor: change file layout and update dependencies

- Move generate, merge, logging, and validate packages to internal/
- Update go.mod dependencies

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

rpkelly committed:
feat: add core SBOM processing functionality

Processor functionality added to enable future merging and deduplication
functionality. This lays the groundwork for sbomtool to read and
manipulate existing sboms

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

ginglis13 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -5,22 +5,89 @@ import (
+
+// Version information
+const (
+	Version   = "1.0.0"
+	BuildDate = "2025-08-04"
+	GitCommit = "dev"
```
I'd prefer this be maintained in its own file (see how containerd does it https://github.com/containerd/containerd/blob/main/version/version.go) or in the Makefile for the tool. Do we need `BuildDate` and `GitCommit`?

ginglis13 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -5,22 +5,89 @@ import (
+	Suggestion string
+}
+
+func (e *validationError) Error() string {
```
just a nit but it looks like the `Error()` impl for validationError and filesystemError has the same logic regarding the `suggestion` but slightly different implementations

ginglis13 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -29,63 +96,355 @@ It analyzes a build directory to identify all components and dependencies, then
 	}
 
-	rootCmd.PersistentFlags().String("log-level", "info", "Log level (debug, info, warn, error)")
+	rootCmd.PersistentFlags().String("log-level", "info", "Set logging verbosity (debug=verbose, info=standard, warn=important, error=critical)")
```
nit but I like the original better.

ginglis13 reviewed (commented):
Just a quick first pass before diving into the core of this PR

arnaldo2792 commented on sbomtool/Makefile:
```diff
@@ -0,0 +1,64 @@
+	@if command -v golangci-lint >/dev/null 2>&1; then \
+		golangci-lint run --timeout=5m; \
+	else \
+		echo "golangci-lint not found, falling back to go vet..."; \
```
Shouldn't this echo be @echo?

arnaldo2792 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+
+	keyToPackages := make(map[string][]*pkg.Package)
+
+	// Group packages by canonical key
```
The extra comments aren't really needed, the code is explicit enough about what's happening.

arnaldo2792 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// getMatchType determines the type of match used for deduplication.
+func (d *Deduplicator) getMatchType(p pkg.Package) string {
```
This could be a standalone function, it doesn't reference `d` from within.

arnaldo2792 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// selectCanonicalPackage chooses the canonical package from a group of duplicates.
+func (d *Deduplicator) selectCanonicalPackage(packages []*pkg.Package) *pkg.Package {
```
This also can be an standalone function.

arnaldo2792 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// calculatePackageCompleteness scores package completeness for canonical selection.
+func (d *Deduplicator) calculatePackageCompleteness(p pkg.Package) int {
```
Same as elsewhere, this can be standalone.

arnaldo2792 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+	return &Deduplicator{config: config}
+}
+
+// DeduplicatePackages performs deduplication on a slice of packages.
```
It seems like result will contain packages by different mappings (Build ID, CPE), along with Canonical Packages, but if you could provide an example of what was the input and what the result after dedup that will be great.

rpkelly replied:
Adding an integ test that demonstrates

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+
+// DeduplicationEvent represents a single event during deduplication processing.
+type DeduplicationEvent struct {
+	Type       string // "cpe_match", "fallback_match", "packages_merged"
```
Could this be an `iota` (Go equivalent of enum) like

```
type DeduplicationEventType int

const (
	cpeMatch DeduplicationEventType = iota // 0
	fallbackMatch          // 1
	packagesMerged        // 2
)
```

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+type DeduplicationConfig struct {
+	PreferCPE        bool   // Prioritize CPE-based deduplication
+	CPENormalization bool   // Apply CPE normalization rules
+	FallbackStrategy string // "name_version_type" or "strict_cpe_only"
```
same as below, can this be an `iota`?

arnaldo2792 replied:
During the offline review, I think this was flagged as "really not needed".

rpkelly commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+	score += len(p.CPEs) * 3                    // CPEs are very important
+	if p.Version != "" {
+		score += 5 // Version info
+	}
```
Fix this to look at actual content

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+	// Note: Files field may not be available in all Syft versions
+	// score += len(p.Files) * 2    // Files are important
+	score += len(p.Licenses.ToUnorderedSlice()) // License info
+	score += len(p.CPEs) * 3                    // CPEs are very important
```
If we have 2 packages:

```
p1 {
  license: Apache
  cpe: ""
}

p2 {
  license: Unknown
  cpe: "cpe:2.3:...."
}
```

We'd end up with `p2`; the license information would be incomplete.

rpkelly commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+	CPENormalization bool   // Apply CPE normalization rules
+	FallbackStrategy string // "name_version_type" or "strict_cpe_only"
+	MetadataMerging  bool   // Enable metadata consolidation
+	PreserveSources  bool   // Track original package sources
```
Take this out

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+				result.Statistics.FallbackMatches++
+			}
+
+			result.ProcessingLog = append(result.ProcessingLog, event)
```
We aren't actually logging the processing log (at least in this commit)

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// normalizeVersion normalizes version strings for consistent matching.
+func (d *Deduplicator) normalizeVersion(version string) string {
```
this doesn't need to be a method on `Deduplicator`

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// mergeLicenses merges license information from duplicate packages.
+func (d *Deduplicator) mergeLicenses(canonical *pkg.Package, duplicates []*pkg.Package) {
```
This isn't calling any methods or accessing any members on `Deduplicator`

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+}
+
+// updatePackageReference updates a package reference with a new canonical ID.
+func (d *Deduplicator) updatePackageReference(ref artifact.Identifiable, canonicalID string) artifact.Identifiable {
```
Is this method needed? Are we stubbing this to be expanded later?

ginglis13 reviewed (commented):
Review of [feat: add deduplication package for SBOM processing](https://github.com/bottlerocket-os/bottlerocket-sdk/pull/297/commits/197ed145234548015285aa12ca4750bc1a848edd)

rpkelly commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,346 @@
+	normalized := strings.TrimPrefix(cpe, "cpe:")
+
+	// Ensure consistent format (cpe:2.3 standard)
+	if !strings.HasPrefix(normalized, "2.3:") {
```
Double check this 2.3

rpkelly replied:
Removing this altogether

ginglis13 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,210 @@
+}
+
+// GetNormalizedPaths extracts normalized paths from scan results for Syft file coordinate matching.
+func (s *Scanner) GetNormalizedPaths(scanResult *ScanResult) []string {
```
this method doesn't access any members or methods of `Scanner`

ginglis13 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,210 @@
+//
+// It removes the buildroot prefix and ensures the path starts with "/" for consistency
+// with Syft's file coordinate system.
+func (n *PathNormalizer) BuildrootToInstalled(buildrootFile, buildrootPath string) (string, error) {
```
Should this method be checking against all of the `buildrootPrefixes` on the `PathNormalizer` ?

ginglis13 commented on sbomtool/internal/processor/processor.go:
```diff
@@ -0,0 +1,140 @@
+
+// GetRelationships returns the relationships from an SBOM for dependency operations.
+// This provides access to Syft's relationship data for filtering operations.
+func (p *SyftConfiguredProcessor) GetRelationships(s *sbom.SBOM) []interface{} {
```
this method isn't accessing members or methods on `SyftConfiguredProcessor`

ginglis13 commented on sbomtool/internal/processor/processor.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+// countPackagesByType counts packages of a specific type for logging purposes.
+func (p *SyftConfiguredProcessor) countPackagesByType(catalog *pkg.Collection, pkgType pkg.Type) int {
```
ditto to above - doesn't need to be a method on `SyftConfiguredProcessor`

ginglis13 commented on sbomtool/internal/validate/validate.go:
```diff
@@ -77,3 +78,126 @@ func ValidateDirectory(path string, purpose string, requireExists bool) error {
+}
+
+// ValidateOutputDirectoryPermissions checks if the output directory is writable.
+func ValidateOutputDirectoryPermissions(outDir string) error {
```
I only see this method being used in `TestValidateOutputDirectoryPermissions`

rpkelly replied:
See CLI commit next

ginglis13 commented on sbomtool/internal/validate/validate.go:
```diff
@@ -77,3 +78,126 @@ func ValidateDirectory(path string, purpose string, requireExists bool) error {
+}
+
+// ValidateOutputPath validates that an output file path is writable.
+func ValidateOutputPath(path string) error {
```
Is this a duplicate of `ValidateOutputPath` in `sbomtool/internal/commands/merge/validator.go` ?
I don't see either method used outside of tests.

ginglis13 commented on sbomtool/internal/validate/validate.go:
```diff
@@ -77,3 +78,126 @@ func ValidateDirectory(path string, purpose string, requireExists bool) error {
 	return nil
 }
+
+// ValidateFilePath validates that a file path exists and is readable.
+func ValidateFilePath(path string, purpose string, mustExist bool) error {
```
Another method I'm not seeing used outside of tests (I'm reviewing commit by commit but for these comments, I'm cross referencing with the entire PR diff)

rpkelly replied:
It's used in the next commit as part of the CLI code

ginglis13 reviewed (commented):
Review of [feat: add core SBOM processing functionality](https://github.com/bottlerocket-os/bottlerocket-sdk/pull/297/commits/988345d3d1a4a734221189a75ac42d2c05c3e464)

ginglis13 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -41,60 +109,342 @@ It analyzes a build directory to identify all components and dependencies, then
+}
+
+// analyzeBuildroot analyzes the buildroot directory and returns a list of files.
+func analyzeBuildroot(buildrootPath string) ([]string, error) {
```
Don't we already get this information from the scanner in `ScanResult` ?

It looks like `analyzeBuildroot` is used to 1/ ensure the buildroot is a directory and 2/ report the number of files inside the buildroot; 2/ seems to be unnecessary as processing later in the program will repeat this effort.

cezar-r commented on sbomtool/README.md:
````diff
@@ -76,12 +84,38 @@ Generate both SPDX and CycloneDX SBOMs:
+
+Merge with debug logging:
+```
+sbomtool --log-level debug merge --output final.json *.json
````
Should we have `*.json` in the example here? The merge command doesn't appear to resolve regex/globbing, so this command won't work for 2 reasons:
1. The `sbomtool` will recognize this as a single file input, which will fail the `validateInputs()` check
2. The command doesn't resolve regex prior to merging, so even without the `validateInputs()` error this will return a `*.json: no such file or directory` error

I'm not sure if the intention is to handle regex/globbing, but if not, we should replace `*.json` here to avoid any confusion

rpkelly replied:
That's a good call out. Usually it's your shell that recognizes globs. I'll remove that for clarity's sake.

arnaldo2792 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -62,9 +62,18 @@ At least one output format (--spdx or --cyclonedx) must be specified.`,
-	generateCmd.MarkFlagRequired("name")
-	generateCmd.MarkFlagRequired("build-dir")
-	generateCmd.MarkFlagRequired("out-dir")
+	if err := generateCmd.MarkFlagRequired("name"); err != nil {
```
For the lack of a better place, and since this is the first commit that violates the message format, the commit should include a brief explanation of why the change is needed. I can see what was done, but during the offline review you explained why the deduplication was required, and why it had to be done following the rules you followed.

arnaldo2792 commented on sbomtool/cmd/sbomtool/root_test.go:
```diff
@@ -122,7 +122,9 @@ func TestLoggingConfiguration(t *testing.T) {
 			// Given: A root command with log level set
 			cmd := createRootCommand()
-			cmd.PersistentFlags().Set("log-level", tt.logLevel)
+			if err := cmd.PersistentFlags().Set("log-level", tt.logLevel); err != nil {
+				t.Fatalf("Failed to set log level: %v", err)
+			}
```
nit: this seems to be more of a cleanup/improvement unrelated to the overall purpose of the PR, you could have a separate commit with improvements to keep the commit isolated from chore changes.

rpkelly replied:
I'm going to take that as a learning for the future.

arnaldo2792 commented on sbomtool/internal/deduplication/deduplicator_bench_test.go:
```diff
@@ -0,0 +1,124 @@
+	"github.com/anchore/syft/syft/pkg"
+)
+
+// BenchmarkDeduplicatePackages_Small benchmarks deduplication with small package sets
```
I think two sets of benchmarks should be enough? One with a big sample size and the other with a smaller size.

arnaldo2792 commented on sbomtool/internal/deduplication/deduplicator_test.go:
```diff
@@ -0,0 +1,502 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewDeduplicator(t *testing.T) {
```
I don't think this test adds too much value, you are already testing this function multiple times in other more complex tests.

arnaldo2792 commented on sbomtool/internal/commands/generate/generate.go:
```diff
@@ -88,7 +88,7 @@ func Generate(name string, spdx bool, cyclonedx bool, buildDir string, outDir st
 	slog.Debug("Creating output directory", "path", outDir)
 	if err := os.MkdirAll(outDir, 0755); err != nil {
-		return false, fmt.Errorf("Failed to create output directory: %w", err)
+		return false, fmt.Errorf("failed to create output directory: %w", err)
```
nit: the changes in this file could also be part of a "cleanup" commit

arnaldo2792 commented on sbomtool/internal/commands/generate/generate_test.go:
```diff
@@ -16,7 +16,11 @@ import (
 	tempFile, err := os.CreateTemp("", "close-test")
 	require.NoError(t, err)
-	defer os.Remove(tempFile.Name())
+	defer func() {
+		if err := os.Remove(tempFile.Name()); err != nil {
+			t.Logf("Failed to remove temp file: %v", err)
+		}
+	}()
```
I'm not Go expert, so I'll lean on @ginglis13 , what's the benefit of changing from the previous code to this new code?

rpkelly replied:
The old code was burying errors

arnaldo2792 commented on sbomtool/internal/logging/logging_test.go:
```diff
@@ -45,40 +46,40 @@ func TestLogLevelConversion(t *testing.T) {
 
 			// The expected level should be enabled
-			if !handler.Enabled(nil, tc.expectedLevel) {
+			if !handler.Enabled(context.TODO(), tc.expectedLevel) {
```
Why is it better to move from `nil` to `context.TODO`?

rpkelly replied:
From the docs: "Do not pass a nil [Context](https://pkg.go.dev/context#Context), even if a function permits it. Pass [context.TODO](https://pkg.go.dev/context#TODO) if you are unsure about which Context to use."

arnaldo2792 commented on sbomtool/pkg/validate/validate.go:
```diff
@@ -37,7 +37,6 @@ func ValidatePackageName(name string) error {
 		}
 	}
 
-
```
nit: one more change for the cleanup commit!

arnaldo2792 commented on sbomtool/internal/validate/validate.go:
```diff
@@ -37,6 +37,7 @@ func ValidatePackageName(name string) error {
 		}
 	}
 
+
```
This space was removed in the previous commit, and re-added here.

arnaldo2792 commented on sbomtool/go.work.sum:
```diff
@@ -119,10 +119,12 @@ cloud.google.com/go/websecurityscanner v1.7.2/go.mod h1:728wF9yz2VCErfBaACA5px2X
 cloud.google.com/go/workflows v1.13.2/go.mod h1:l5Wj2Eibqba4BsADIRzPLaevLmIuYF2W+wfFBkRG3vU=
 github.com/GoogleCloudPlatform/docker-credential-gcr v2.0.5+incompatible/go.mod h1:BB1eHdMLYEFuFdBlRMb0N7YGVdM5s6Pt0njxgvfbGGs=
 github.com/GoogleCloudPlatform/opentelemetry-operations-go/detectors/gcp v1.25.0/go.mod h1:obipzmGjfSjam60XLwGfqUkJsfiheAl+TUjG+4yzyPM=
+github.com/acarl005/stripansi v0.0.0-20180116102854-5a71ef0e047d h1:licZJFw2RwpHMqeKTCYkitsPqHNxTmd4SNR5r94FGM8=
```
Why if this is just a move from `pkg` to `internal` were the dependencies updated?

arnaldo2792 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,210 @@
+// Package buildroot provides utilities for scanning and analyzing buildroot directory structures.
+//
+// The package implements efficient recursive directory traversal with configurable depth limits,
+// file type detection, and path normalization for integration with Syft's file coordinate system.
```
This is a bit repetitive with the next doc comment, I think you can drop it.

arnaldo2792 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,210 @@
+	entries, err := os.ReadDir(currentPath)
+	if err != nil {
+		slog.Warn("Failed to read directory", "path", currentPath, "error", err)
+		return nil // Continue scanning other directories
```
This function also suffers from the "unnecessary comments" feedback.

arnaldo2792 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,210 @@
+// Package buildroot provides utilities for scanning and analyzing buildroot directory structures.
```
Why does the tool have to scan the entire build root instead of just the paths we know have to be scanned? Isn't the sbom created from either a binary (e.g. /sbin, /usr/bin) or a directory with SBOMs that we control?

rpkelly replied:
We talked about this offline but, essentially, we scan the entire build directory when the sources are unpacked and the binary is being built. This gives us a superset of what will be on the image. We then filter at the end of the install step, when we know what's actually on the image, but would not be guaranteed to have access to all of the dependencies if we had just generated then.

arnaldo2792 commented on sbomtool/internal/validate/validate.go:
```diff
@@ -77,3 +78,126 @@ func ValidateDirectory(path string, purpose string, requireExists bool) error {
+}
+
+// DetectSBOMFormat detects the format of an SBOM file by examining its content.
+func DetectSBOMFormat(path string) (string, error) {
```
Isn't the SBOM in JSON? Why not just parse it in a struct and let failed deserialization be the validation?

arnaldo2792 commented on sbomtool/cmd/sbomtool/generate_test.go:
```diff
@@ -26,12 +26,12 @@ func TestGenerateCommandMetadata(t *testing.T) {
 			name:     "correct short description",
 			field:    "Short",
-			expected: "Generate SBOM files for a directory",
+			expected: "Generate comprehensive SBOM files from build directories",
```
I've seen "comprehensive" quite a lot in this change, wasn't the previous SBOM already "comprehensive"? This change doesn't add more to the SBOM if anything it simplifies it.

arnaldo2792 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -115,14 +465,79 @@ func runGenerate(cmd *cobra.Command, args []string) error {
 	if !success {
-		return fmt.Errorf("no SBOM formats were generated")
+		return &validationError{
+			Field:      "SBOM generation result",
+			Value:      "no formats generated",
+			Message:    "no SBOM formats were successfully generated",
+			Suggestion: "Check that the build directory contains recognizable package formats. Supported formats include Go modules, npm packages, Python requirements, and more.",
+		}
+	}
+
+	slog.Info("SBOM generation completed successfully")
+
+	if spdx {
+		spdxFile := filepath.Join(outDir, fmt.Sprintf("%s-spdx.json", name))
+		slog.Info("Generated SPDX SBOM", "file", spdxFile)
+	}
+	if cyclonedx {
+		cyclonedxFile := filepath.Join(outDir, fmt.Sprintf("%s-cyclonedx.json", name))
+		slog.Info("Generated CycloneDX SBOM", "file", cyclonedxFile)
+	}
+
+	return nil
+}
```
This block seems like an unrelated change to the commit(?). The commit integrates the new sbom functionality (merge and dedup), but in here you are combining both the new functionality with more verbose (useful?) error messages. Which again, should have been kept in its own commit.

arnaldo2792 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -115,14 +465,79 @@ func runGenerate(cmd *cobra.Command, args []string) error {
+	}
+	slog.Info("Generating SBOM formats", "formats", formats)
+
+	if err := performPreflightChecks(buildDir, outDir); err != nil {
```
It seems like `performPreflightChecks` performs some duplicate validations that are individually implemented in the next lines.

arnaldo2792 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -156,7 +571,7 @@ func runMerge(cmd *cobra.Command, args []string) error {
 	_, err := merge.Merge(level, args)
 	if err != nil {
-		return fmt.Errorf("SBOM merge failed: %w", err)
+		return fmt.Errorf("sBOM merge failed: %w", err)
```
This should have been kept as SBOM not sBOM

arnaldo2792 commented on sbomtool/cmd/sbomtool/root_test.go:
```diff
@@ -69,7 +69,7 @@ func TestPersistentFlags(t *testing.T) {
 		require.NotNil(t, flag, "log-level flag should be registered")
 		assert.Equal(t, "info", flag.DefValue, "Default log level should be 'info'")
-		assert.Equal(t, "Log level (debug, info, warn, error)", flag.Usage, "Flag usage should be descriptive")
+		assert.Equal(t, "Set logging verbosity (debug=verbose, info=standard, warn=important, error=critical)", flag.Usage, "Flag usage should be descriptive")
```
I wouldn't bother about these type of unit tests. They will be a pain to maintain and don't actually prevent any unwanted behavior.

arnaldo2792 commented on sbomtool/internal/commands/filter/filtering.go:
```diff
@@ -0,0 +1,371 @@
+func NewFilteringEngine() *FilteringEngine {
+	return &FilteringEngine{
+		processor:        processor.NewBottlerocketSyftProcessor(),
+		buildrootScanner: buildroot.NewScanner(10, []string{"*.tmp", "*.log"}),
```
Why is 10 the correct depth?

rpkelly replied:
switching to 30. I've looked at some Bottlerocket images and they have a depth of about 15, so, 30 should be enough to get everything, but still guard against any sort of loops.

arnaldo2792 commented on sbomtool/internal/commands/filter/filtering.go:
```diff
@@ -0,0 +1,371 @@
+}
+
+// FilterSBOMByBuildroot filters an SBOM based on buildroot contents with transitive dependency resolution.
+func (e *FilteringEngine) FilterSBOMByBuildroot(
```
Most of the functions in FilteringEngine could be stand alone functions.

arnaldo2792 commented on sbomtool/README.md:
```diff
@@ -76,12 +84,38 @@ Generate both SPDX and CycloneDX SBOMs:
 
+## Deduplication Behavior
+
+The merge command uses intelligent deduplication to combine packages from multiple SBOMs:
```
`intelligent` seems like a stretch. I would just say `deduplication` and that's it.

arnaldo2792 commented on sbomtool/README.md:
```diff
@@ -76,12 +84,38 @@ Generate both SPDX and CycloneDX SBOMs:
+The merge command uses intelligent deduplication to combine packages from multiple SBOMs:
+
+### CPE-Based Deduplication
+- **Primary Strategy**: Uses Common Platform Enumeration (CPE) as the canonical identifier
```
You have already mentioned `CPE` earlier in the README, I think you can just use `CPE` here.

arnaldo2792 commented on sbomtool/internal/commands/merge/validator.go:
```diff
@@ -0,0 +1,82 @@
+// Package merge provides validation functionality for merge operations.
```
In other commits, the validations are performed by individual functions rather than by struct. For consistency, you should choose one approach (either an interface with implementations, or just plain functions)

arnaldo2792 commented on sbomtool/internal/processor/processor.go:
```diff
@@ -81,6 +81,11 @@ func (p *SyftConfiguredProcessor) LoadSBOM(path string) (*sbom.SBOM, string, err
 	return s, string(formatID), nil
 }
 
+// GetFormatEncoders returns the configured format encoders
+func (p *SyftConfiguredProcessor) GetFormatEncoders() []sbom.FormatEncoder {
+	return p.formatEncoders
+}
+
```
Why was this moved? Didn't it work where it was before?

rpkelly committed:
feat: Add filter module to enable buildroot filtering

During initial SBOM creation all sources in the initial build are added
to the SBOM, however, many of these sources are never actually built or
added to the resulting Bottlerocket rpm. Buildroot filtering takes the
initial large SBOM as well as the buildroot of actually installed files
to limit the packages contained in the SBOM to only those we care about

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

rpkelly committed:
feat: integrate new functionality with CLI

Add filtering and deduplication calls to the cli, allowing tool to
remove extra and unnecessary information from generated SBOMs

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

ginglis13 commented on sbomtool/Makefile:
```diff
@@ -0,0 +1,64 @@
+	@echo "Running tests..."
+	@go work sync
+	@go test $$(go list ./... | grep -v "/pkg/config")
+	@echo "Tests complete"
```
I don't see any config pkg:
```suggestion
# Run tests
test:
	@echo "Running tests..."
	@go work sync
	@go test ./...
	@echo "Tests complete"
````

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,304 @@
+// DeduplicationResult contains the results of package deduplication.
+type DeduplicationResult struct {
+	CanonicalPackages map[string]*pkg.Package // canonical key -> package
+	IDMapping         map[string]string       // old package ID -> canonical package ID
```
What is "old package ID" ? From my read this is a kind of hash generated by `syft/syft/pkg/package.go` `SetID` method? https://github.com/anchore/syft/blob/9217f2099f57f1376fd4a6df62ce6c68d105d0cd/syft/pkg/package.go#L37

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,304 @@
+	version := normalizeVersion(p.Version)
+	pkgType := string(p.Type)
+
+	return fmt.Sprintf("fallback:%s|%s|%s", name, version, pkgType)
```
nit but I would prefer we just keep this as 

```
%s|%s|%s
```

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,304 @@
+	return normalized == "unknown" || normalized == "noassertion" || normalized == ""
+}
+
+// mergeOverlappingCPEGroups merges package groups that have overlapping CPEs
+func mergeOverlappingCPEGroups(keyToPackages map[string][]*pkg.Package) map[string][]*pkg.Package {
```
similar to a comment from @arnaldo2792 , it would be nice to have a unit test for this method both for test coverage and as an example.

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,304 @@
+		}
+	}
+	if !foundValidVersion {
+		for _, p := range packages {
+			if p.Version != "" {
+				canonical.Version = p.Version
+				break
+			}
+		}
```
is this iteration necessary? Based on the comment and the conditional in the previous iteration, can't you skip this and just set `canonical.Version` to unknown?

rpkelly replied:
The goal is to get the first non-empty string descriptor used among the merged packages (either "Unknown" or "NOASSERTION" in some capitaliztion scheme). That's not necessarily the very first one, and it's not necessarily unknown either.

ginglis13 reviewed (commented):
another pass on [feat: add deduplication package for SBOM processing](https://github.com/bottlerocket-os/bottlerocket-sdk/pull/297/commits/36bf08e7f93e28ac402a9b533e71a37ea3502b95)

ginglis13 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,186 @@
+
+	// Normalize all paths for Syft compatibility
+	for _, filePath := range result.AllFiles {
+		normalized, err := BuildrootToInstalled(filePath, buildrootPath)
```
nit but IMO this would be better as a method on a `buildrootPath` type to avoid the footgun of mis-ordering these method arguments:

```
buildrootPath.InstalledPath(filepath)
```

or even model the `buildrootPath` on the `Scanner` and call as

```
s.BuildrootToInstalled(filePath)
```

ginglis13 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,186 @@
+
+		case info.IsDir():
+			result.Directories = append(result.Directories, fullPath)
+			// Recurse into subdirectory
```
unnecessary comment

ginglis13 commented on sbomtool/internal/processor/processor.go:
```diff
@@ -0,0 +1,119 @@
+// GenerateComprehensiveSBOM generates an SBOM using configured Syft catalogers.
+// It includes comprehensive package detection with Go and Rust binary analysis.
+func GenerateComprehensiveSBOM(buildDir string) (*sbom.SBOM, error) {
+	// Create source from directory
```
comment doesn't seem necessary - maybe "Create a syft source object that syft can scan for package discovery and SBOM generation"

ginglis13 commented on sbomtool/internal/processor/processor.go:
```diff
@@ -0,0 +1,119 @@
+
+// GetRelationships returns the relationships from an SBOM for dependency operations.
+// This provides access to Syft's relationship data for filtering operations.
+func GetRelationships(s *sbom.SBOM) []interface{} {
```
I only see this method being used in tests

ginglis13 reviewed (commented):
another pass of [feat: add core SBOM processing functionality](https://github.com/bottlerocket-os/bottlerocket-sdk/pull/297/commits/0e2af9f4f52c266469a6fb81d2b8690c3820efb6)

arnaldo2792 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,182 @@
+		}
+		if matched, _ := filepath.Match(pattern, filepath.Base(path)); matched {
+			return true
+		}
```
In here you are silently ignoring a potential error, if the pattern (which is a configurable input) isn't a valid pattern. You should either validate the patterns in `NewScanner` or handle the error here.

arnaldo2792 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,182 @@
+	}
+
+	// Clean the path for Syft compatibility
+	installedPath := filepath.Clean(relativePath)
```
nit: a couple of more unnecessary comments.

arnaldo2792 commented on sbomtool/internal/processor/processor_test.go:
```diff
@@ -0,0 +1,258 @@
+func TestProcessorIntegration(t *testing.T) {
+	// GIVEN: A realistic test scenario
+	// WHEN: Multiple operations are performed in sequence
+	// THEN: All operations should work together correctly
```
These are too generic descriptions, the description should be more concrete on what the test is actually testing.

arnaldo2792 commented on sbomtool/internal/processor/processor_test.go:
```diff
@@ -0,0 +1,258 @@
+		require.NoError(t, err)
+	}
+
+	// Create test files
```
Unnecessary comments

arnaldo2792 commented on sbomtool/internal/validate/validate.go:
```diff
@@ -76,3 +79,111 @@ func ValidateDirectory(path string, purpose string, requireExists bool) error {
 }
+
+// ValidateFilePath validates that a file path exists and is readable.
+func ValidateFilePath(path string, purpose string, mustExist bool) error {
```
I think this and the next functions are a bit extra. If either reads or writes fail, you will get an error from standard Go explaining what failed in the operation, so you really don't have to spend extra IO checking the files permissions. The same with the other Validation performed in later, for the output directory.

rpkelly replied:
For the output I actually think it's probably worth it since generating the sbom can be expensive, and it's probably better to validate where it's going at the beginning rather then after we've generated it.

arnaldo2792 commented on sbomtool/internal/validate/validate.go:
```diff
@@ -76,3 +79,111 @@ func ValidateDirectory(path string, purpose string, requireExists bool) error {
+		return "", nil
+	}
+
+	return string(formatID), nil
```
Do you have to wrap `formatID` in `string()`? It seems like `formatID` is already a string. Even if it isn't, you can just call `formatID.String()`:

https://github.com/anchore/syft/blob/b9604cbf3028d319a18b728a3f36a37767fb254d/syft/sbom/format.go#L10

arnaldo2792 commented on sbomtool/internal/commands/filter/filtering.go:
```diff
@@ -0,0 +1,351 @@
+		"buildroot_path", buildrootPath)
+
+	// Bottlerocket is about 15 layers deep, so 30 will get everything but prevents infinite loops
+	buildrootScanner := buildroot.NewScanner(30, []string{"*.tmp", "*.log"})
```
The Bottlerocket root images won't have any of these type of files. It will however include some files that you may not care about for SBOM purposes, e.g. all the kernel modules, the kernel headers, licenses, the application inventory, etc.

rpkelly replied:
That's true, but they don't hurt. I can remove them if you really want though

arnaldo2792 commented on sbomtool/internal/commands/filter/filtering.go:
```diff
@@ -0,0 +1,351 @@
+		return nil, fmt.Errorf("filtered SBOM validation failed: %w", err)
+	}
+
+	allPackages := inputSBOM.Artifacts.Packages.Sorted()
```
Does ordering matter here?

ginglis13 replied:
is this sort necessary? in `getExcludedPackages`, we compare against a set, so the order of items in this array shouldn't impact that comparison to construct the `excludedPackages`. If `excludedPackages` must be sorted (which it is as a result of pre-sorting `allPackages`), I'd prefer to explicitly sort it instead.

arnaldo2792 commented on sbomtool/internal/commands/filter/filtering.go:
```diff
@@ -0,0 +1,351 @@
+
+	var rootPackages []pkg.Package
+
+	for _, p := range packages.Sorted() {
```
Is the outcome of this function impacted if the packages aren't sorted?

arnaldo2792 commented on sbomtool/internal/commands/filter/filtering.go:
```diff
@@ -0,0 +1,351 @@
+		if relType == depType {
+			return true
+		}
+	}
+	return false
+}
```
You should be able to use `slices.Contains` if you make RELATIONSHIP_TYPES a const:

```golang
slices.Contains(RELATIONSHIP_TYPES, relType)
```

arnaldo2792 commented on sbomtool/internal/commands/filter/filtering.go:
```diff
@@ -0,0 +1,351 @@
+
+// SaveFilteredSBOM saves the filtered SBOM to a file using the specified format.
+// It leverages the processor's format handling capabilities to maintain format compatibility.
+func SaveFilteredSBOM(
```
I pulled down the code, and grep for this. This function is only used in the test that was written for it:

```
└─> on  sbomtool-features ✦ ❯ rg SaveFiltered
sbomtool/internal/commands/filter/filtering_test.go
463:func TestSaveFilteredSBOM(t *testing.T) {
465:	// WHEN: SaveFilteredSBOM is called
479:	err := SaveFilteredSBOM(result, outputPath, "syft-json")

sbomtool/internal/commands/filter/filtering.go
314:// SaveFilteredSBOM saves the filtered SBOM to a file using the specified format.
316:func SaveFilteredSBOM(
```

arnaldo2792 commented on sbomtool/cmd/sbomtool/merge_integration_test.go:
```diff
@@ -0,0 +1,214 @@
+	require.NoError(t, err, "Direct merge should work")
+
+	// Test the in-memory SBOM packages
+	packages := make([]pkg.Package, 0)
+	for p := range result.MergedSBOM.Artifacts.Packages.Enumerate() {
+		packages = append(packages, p)
+	}
```
Why do you need to create another list for the packages if `Packages` already has the packages that you are testing?

rpkelly replied:
This is a test, and the slice is the easiest way to be able to call `len`

arnaldo2792 commented on sbomtool/cmd/sbomtool/merge_integration_test.go:
```diff
@@ -0,0 +1,214 @@
+	// Test the in-memory SBOM directly (like deduplication tests)
+	// Re-run the merge logic to get the in-memory result
+	config := merge.MergeConfig{
+		OutputPath: outputFile,
+		Level:      0,
+	}
+	result, err := merge.Merge(config, []string{file1, file2})
```
I just realized by reading this  test that probably `Merge` shouldn't write the file, and this test made it more obvious. If you don't write in `Merge` and just let it do the merge, you don't have to do the dance to create a file or ignore it after you call the `Merge` function.

rpkelly replied:
the actual file write has been moved to `main.go`

arnaldo2792 commented on sbomtool/cmd/sbomtool/merge_integration_test.go:
```diff
@@ -0,0 +1,214 @@
+	assert.Equal(t, 5, len(packages), "Should have 5 packages: 2 app components + 3 deduplicated libraries")
+
+	// Verify specific library packages exist (the actual dependencies)
+	packageNames := make([]string, len(packages))
+	for i, p := range packages {
+		packageNames[i] = p.Name
+	}
+	assert.Contains(t, packageNames, "pkg1")
+	assert.Contains(t, packageNames, "pkg2")
+	assert.Contains(t, packageNames, "pkg3")
+	assert.Contains(t, packageNames, "app1") // Metadata component
+	assert.Contains(t, packageNames, "app2") // Metadata component
+}
```
I was thinking there is a way to write this in a more idiomatic way, but realized that it will loose the context of what item actually is missing.

ginglis13 commented on sbomtool/internal/deduplication/deduplicator.go:
```diff
@@ -0,0 +1,300 @@
+	// Use non-empty version if available, otherwise keep unknown if all are unknown
+	foundValidVersion := false
+	for _, p := range packages {
+		if p.Version != "" && !isUnknownOrNoAssertion(p.Version) {
```
looks like `!= ""` here is redundant because that cause is also covered by `isUnknownOrNoAssertion`

rpkelly replied:
`isUnknownOrNoAssertion` allows empty string so here we're checking that it's not an empty string.

arnaldo2792 commented on sbomtool/cmd/sbomtool/merge_test.go:
```diff
@@ -6,13 +6,61 @@
 	"github.com/stretchr/testify/require"
 )
 
+func TestMergeCommandMetadata(t *testing.T) {
```
I wouldn't bother with this test, as it only tests the messages for the usage, and not the actual behavior of the test.

ginglis13 commented on sbomtool/internal/buildroot/scanner.go:
```diff
@@ -0,0 +1,182 @@
+	}
+
+	entries, err := os.ReadDir(currentPath)
+	if err != nil {
+		slog.Warn("Failed to read directory", "path", currentPath, "error", err)
+		return nil
+	}
```
Why doesn't this `return err` ?

arnaldo2792 commented on sbomtool/cmd/sbomtool/main.go:
```diff
@@ -447,32 +447,71 @@ func createMergeCommand() *cobra.Command {
 	}
 
-	mergeCmd.Flags().Int("level", 0, "Merge level")
+	mergeCmd.Flags().String("output", "", "Output file path for merged SBOM (required)")
+	mergeCmd.Flags().Int("level", 0, "Merge level (reserved for future use)")
+	if err := mergeCmd.MarkFlagRequired("output"); err != nil {
+		slog.Error("Failed to mark output flag as required", "error", err)
+		os.Exit(1)
+	}
```
I'm not a Cobra expert, will Cobra prevent calling `merge` without any input files? If it doesn't, then the code that takes the first input as the base will throw an out of bounds error.

rpkelly replied:
That's handled by the `cobra.MinimumNArgs(2)` call above

arnaldo2792 commented on sbomtool/cmd/sbomtool/merge_integration_test.go:
```diff
@@ -0,0 +1,214 @@
+
+	// Test the in-memory SBOM packages
+	packages := make([]pkg.Package, 0)
+	for p := range result.MergedSBOM.Artifacts.Packages.Enumerate() {
+		packages = append(packages, p)
+	}
```
Same as elsewhere, you don't really need `Packages` as a new slice.

rpkelly replied:
Same as above, the slice makes the test much easier to deal with.

arnaldo2792 commented on sbomtool/internal/commands/merge/merge.go:
```diff
@@ -1,17 +1,195 @@
+		return &sbom.SBOM{}, nil, nil
+	}
+
+	// Use first SBOM as base
+	merged := &sbom.SBOM{
+		Descriptor: sboms[0].Descriptor,
+		Source:     sboms[0].Source,
+	}
+
```
This is where you may get an out of bounds if the sboms were empty.

rpkelly replied:
This is checked on the first line of the function.

ginglis13 commented on sbomtool/Makefile:
```diff
@@ -0,0 +1,64 @@
+	@go build -o $(BINARY_NAME) $(CMD_DIR)
+	@echo "Build complete: $(BINARY_NAME)"
+
+# Run tests (excluding broken config package)
```
nit: this comment should be updated as there is no test being excluded. IMO however, most of the comments above individual Makefile rules are unnecessary

ginglis13 commented on sbomtool/internal/processor/processor_test.go:
```diff
@@ -28,43 +26,6 @@ func TestNewBottlerocketSyftProcessor(t *testing.T) {
 	assert.Greater(t, len(encoders), 0)
 }
 
-func TestGenerateComprehensiveSBOM(t *testing.T) {
```
What are these changes doing in this commit?

ginglis13 replied:
It looks like these changes to `sbomtool/internal/processor` belong in [feat: add core SBOM processing functionality](https://github.com/bottlerocket-os/bottlerocket-sdk/pull/297/commits/6ba1bbfd7efc9cfc5271831c943ef2bd384ebb9e)

ginglis13 commented:
Just a couple of nits, otherwise LGTM
